### DISCUSSION TITLE
Optional Body(...) parameter?

### DISCUSSION THREAD

**Discussion by charliegriefer** [fastapi/fastapi#8189] [Questions] [answered]
Hey folks... 

I have an API request (a POST) that currently takes one path var and two body vars as follows:

```
@api.post("/account/{account_id}/report", status_code=201)
def create_report(account_id: int, user_params: UserParams, report_type: str = Body(...)):
```
(`UserParams` is a pydantic BaseModel defined earlier in the script)

I now need to add another body param. Just a string. But it's optional. 

I didn't see anything in the FastAPI docs that demonstrated an optional Body param, other than setting a BaseModel value to `None`.

I looked at the source code and saw that `Body` takes a `Schema`, but haven't been able to figure out how to make it optional or default it to `None`.

I tried:

```
def create_report(account_id: int, user_params: UserParams, report_type: str = Body(...), sf_rec_id: int = Body(..., default=None)):
```
But got the following exception:

`TypeError: Body() got multiple values for argument 'default'`

Any help would be greatly appreciated.

Thanks!

**Comment by charliegriefer** [comment]
_sigh_

Nevermind. Looks like this does it:

```
def create_report(account_id: int, user_params: UserParams, report_type: str = Body(...), sf_rec_id: str = Body(default=None)):
```

Out of curiosity, would anybody mind elaborating on the `...` that gets passed in? It looks like I don't need to use it if I'm sending in a default value. I tried the other way because I saw `importance: int = Body(..., gt=0),` in the FastAPI docs, so assumed that the `...` was necessary and any other args would follow. Clearly, that was incorrect :) 

Thanks!

**Comment by tiangolo** [comment, marked as answer]
Thanks for reporting back and closing the issue.

The `...` means "this is required", in contrast to `None`, not required, or a literal value `"Foo"`, that means that `"Foo"` is the default.

**Comment by jibinmathew69** [comment]
You changed the datatype in the answer compared to the Question, but what if the optional data is supposed to be int/float and default value is None.

**Comment by igorbga** [comment]
@jibinmathew69 I don't understand your question.
If your default value is actually "None", then your type is not int/float, is Optional[int/float]. "..." would mean that the value is required meanwhile None would mean that actually your default value is None. 

Of course this means you can not distinguise when a "myvalue: null" was sent and when directly "myvalue" was not set. I think this problem is quite extended when processing JSON input.

Pydantic models "solve" it by having control over unset values (https://fastapi.tiangolo.com/tutorial/body-updates/#partial-updates-with-patch), but I guess for "primitive" types you don't have such options.

Was that your question?

**Comment by bostrick** [comment]
I'm having similar issues... I'd like to define a body as either an arbitrary dict or just no body... I've tried variants such as... 

```
ActionCommandParamsT = Annotated[dict[str, Any] |None, Body()]
ActionCommandParamsT = Annotated[dict[str, Any], Body()] | None
ActionCommandParamsT = Annotated[dict[str, Any], Body(default=None)]
```

but none seem to do the magic (the first two causing route confusion between Body() and Query() params, the latter just plain wrong.)

Anyone know some magic I"m missing?    Thanks!
